pydantic_ai_slim/pydantic_ai/models/gemini.py

```diff
@@ -241,32 +241,32 @@ async def _process_streamed_response(http_response: HTTPResponse) -> EitherStrea
         if start_response is None:
             raise UnexpectedModelBehavior('Streamed response ended without content or tool calls')
 
+        # TODO: Update this once we rework stream responses to be more flexible
         if _extract_response_parts(start_response).is_left():
             return GeminiStreamStructuredResponse(_content=content, _stream=aiter_bytes)
         else:
             return GeminiStreamTextResponse(_json_content=content, _stream=aiter_bytes)
 
     @staticmethod
-    def _message_to_gemini(m: Message) -> _utils.Either[_GeminiTextPart, _GeminiContent]:
+    def _message_to_gemini_system_prompt(m: Message) -> _GeminiTextPart | None:
         """Convert a message to a _GeminiTextPart for "system_instructions" or _GeminiContent for "contents"."""
-        if m.role == 'system':
-            # SystemPrompt ->
-            return _utils.Either(left=_GeminiTextPart(text=m.content))
-        elif m.role == 'user':
-            # UserPrompt ->
-            return _utils.Either(right=_content_user_text(m.content))
-        elif m.role == 'tool-return':
-            # ToolReturn ->
-            return _utils.Either(right=_content_function_return(m))
-        elif m.role == 'retry-prompt':
-            # RetryPrompt ->
-            return _utils.Either(right=_content_function_retry(m))
-        elif m.role == 'model-text-response':
-            # ModelTextResponse ->
-            return _utils.Either(right=_content_model_text(m.content))
-        elif m.role == 'model-structured-response':
-            # ModelStructuredResponse ->
-            return _utils.Either(right=_content_function_call(m))
+        if isinstance(m, SystemPrompt):
+            return _GeminiTextPart(text=m.content)
+        else:
+            return None
+
+    @staticmethod
+    def _message_to_gemini_content(m: Message) -> _GeminiContent | None:
+        if isinstance(m, SystemPrompt):
+            return None
+        elif isinstance(m, UserPrompt):
+            return _content_user_prompt(m)
+        elif isinstance(m, ToolReturn):
+            return _content_tool_return(m)
+        elif isinstance(m, RetryPrompt):
+            return _content_retry_prompt(m)
+        elif isinstance(m, ModelResponse):  # type: ignore[reportUnnecessaryIsInstance]
+            return _content_model_response(m)
```

Comment:
Note — here and elsewhere I have replaced checks on `m.role` with `isinstance` checks on `m`. This makes it easier to rename things — my IDE won't automatically update the values of `role` (yes I can find-replace, but I don't love assuming that there aren't unrelated usages of that string), and I can't as easily do a cmd+click jump to the definition of the relevant type if we just check the discriminator to narrow it.

If there was a good reason to do it with the role value checking we can go back to that, but if it was just performance relative to an `isinstance`, I personally think the codebase ergonomics benefits are worth more than the nearly-insignificant performance cost.